### PR TITLE
fix: Update filtering of STX for the Activity list

### PR DESCRIPTION
**Pull request by dan437**
## **Description**

Update filtering of STX for the Activity list. There was a rare edge case when we showed a duplicated transaction in the Activity list when the transaction status was "unknown" due to reverting on chain. Now backend returns "reverted" for such transactions instead of "unknown", but just in case we have updated client-side code to only show "unknown" non-swaps transactions once in the Activity list.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/24540?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Have at least one smart transaction submitted e.g. on Sepolia
2. In the `smartTransactionsListSelector` function you can artificially add a new smart transaction with the status "unknown". It should not show up in the Activity list.
```
const smartTransactions =
    state.metamask.smartTransactionsState?.smartTransactions?.[
      getCurrentChainId(state)
    ];
const smartTransactionsCopy = [...smartTransactions];
smartTransactionsCopy.push({
  ...smartTransactions[0],
  status: 'unknown',
  type: 'contractInteraction',
  statusMetadata: {
    ...smartTransactions[0].statusMetadata,
    minedTx: 'unknown',
  },
});
```

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Review by antonydenyer — DISMISSED** [review]

File: ui/selectors/transactions.js [hunk 1]
```diff
@@ -20,6 +21,12 @@ const INVALID_INITIAL_TRANSACTION_TYPES = [
   TransactionType.cancel,
   TransactionType.retry,
 ];
+const allowedSwapsSmartTransactionStatusesForActivityList = [
+  SmartTransactionStatuses.PENDING,
+  SmartTransactionStatuses.UNKNOWN,
+  SmartTransactionStatuses.RESOLVED,
+  SmartTransactionStatuses.CANCELLED,
```
Comment (dbrans) [inline ui/selectors/transactions.js:28]: Hey Dan! Could you include a comment explaining why this particular subset of statuses was chosen? In particular, why we omit the `SUCCESS` and all the `CANCELLED_X` statuses?

File: ui/selectors/transactions.js [hunk 2]
```diff
@@ -102,18 +109,25 @@ export const smartTransactionsListSelector = (state) => {
   return state.metamask.smartTransactionsState?.smartTransactions?.[
     getCurrentChainId(state)
   ]
-    ?.filter((stx) => {
-      const isCancelledSmartTransaction = stx.status?.startsWith('cancelled');
+    ?.filter((smartTransaction) => {
```
Comment (dbrans) [inline ui/selectors/transactions.js:112]: Quick note unrelated to this PR: Should we add a comment explaining that we only show swap smart transactions in the activity list and why?

File: ui/selectors/transactions.test.js [hunk 3]
```diff
@@ -128,6 +136,255 @@ describe('Transaction Selectors', () => {
       expect(msgSelector).toStrictEqual([msg]);
     });
   });
+  describe('smartTransactionsListSelector', () => {
```
Comment (dbrans) [inline ui/selectors/transactions.test.js:139]: Great tests ❤️

File: ui/selectors/transactions.js (hunk 2 above)
Comment (dan437) [reply ui/selectors/transactions.js:120]: We also show pending non-swap smart transactions in the Activity list. I've added a few comments to make it a bit clearer.

File: ui/selectors/transactions.js [hunk 4]
```diff
@@ -21,6 +22,20 @@ const INVALID_INITIAL_TRANSACTION_TYPES = [
...
+// The SUCCESS status is not allowed because once a Smart Swap is successful,
+// the "this.confirmExternalTransaction" function is called, which moves the transaction
+// to the regular transaction list in the TransactionController.
+// TODO: When Swaps and non-Swaps transactions are treated the same,
+// we will only allow the PENDING smart transaction status in the Activity list.
+const allowedSwapsSmartTransactionStatusesForActivityList = [
+  SmartTransactionStatuses.PENDING,
+  SmartTransactionStatuses.UNKNOWN,
+  SmartTransactionStatuses.RESOLVED,
+  SmartTransactionStatuses.CANCELLED, // Other CANCELLED_ statuses are deprecated.
```
Comment (dbrans) [inline ui/selectors/transactions.js:36]: Instead of putting it here, let's add the comment about deprecated statuses to the original enum in the smart transaction controller repo. We can do this in a separate PR.

File: ui/selectors/transactions.js [hunk 5]
```diff
@@ -21,6 +22,20 @@ const INVALID_INITIAL_TRANSACTION_TYPES = [
   TransactionType.retry,
 ];
 
+// The statuses listed below are permitted in the Activity list for Smart Swaps because
+// a Smart Swap with these statuses hasn't been added to the TransactionController's regular transaction list yet.
+// The SUCCESS status is not allowed because once a Smart Swap is successful,
+// the "this.confirmExternalTransaction" function is called, which moves the transaction
+// to the regular transaction list in the TransactionController.
```
Comment (dbrans) [inline ui/selectors/transactions.js:29]: Hey, I believe we can simplify this a bit to make it clearer. What are your thoughts?
```suggestion
// The statuses listed below are allowed in the Activity list for Smart Swaps
// The SUCCESS and REVERTED statuses are excluded because smart transactions with 
// those statuses are already in the regular transaction list.
```

**Review by dbrans — DISMISSED** [review]

**Review by antonydenyer — APPROVED** [review]

**Review by dbrans — APPROVED** [review]
